### PR TITLE
Add branch selection

### PR DESCRIPTION
**Pull request by TShapinsky**
- Makes branch name selectable
- Branch list includes local and remote branches and when selected fetches remotes
- If there are unsaved changes it prompts the user who can choose whether they wish to checkout the new branch causing a `git stash` followed by a `git checkout`

demo

### CODE REVIEW

File: lib/branch-list-view.coffee [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+  initialize: ->
+    super
+
+    emitter = new Emitter
+    emitter.on 'on-change-branch', (branch) ->
+      atom.workspace.getActivePaneItem().save()
+    @panel = atom.workspace.addModalPanel(item: this, visible: false)
+    @addClass('branch-selector')
+    @list.addClass('mark-active')
+
```
Comment (mnquintana) [inline lib/branch-list-view.coffee]: :art: There should only be 1 extra newline here

File: lib/branch-list-view.coffee [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+                    exit: (code) ->
+                      git
+                        args: ['checkout', branch.full]
+                        cwd: cwd
+                        exit: (code) ->
+                          emitter.emit 'on-change-branch', branch.name}]
+              dismissable: true
+            notification = atom.notifications.addError("You have uncommitted changes", options)
+        else
+          emitter.emit 'on-change-branch', branch.name
```
Comment (mnquintana) [inline lib/branch-list-view.coffee]: :art: Can you add a newline here?

**Comment by thedaniel** [comment]
Real quick drive-by feedback: you shouldn't be using a red notification box unless it's an exception / error state in the editor or a package.

**Comment by thedaniel** [comment]
Also, I wonder if this shouldn't be a service provided by a separate package to make it pluggable for alternate VCS.

File: lib/branch-list-view.coffee [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+              name = full.replace("remotes/", "")
+              remote = branch.includes("remotes/")
+              current = branch.startsWith("* ")
+              unless full.includes("->")
+                branches.push({name: name, remote: remote, current: current, full: full})
+
+  getFilterKey: ->
+    'name'
+
+  git = ({args, cwd, options, stdout, stderr, exit}={}) ->
```
Comment (benogle) [inline lib/branch-list-view.coffee]: This is weird in the middle of the class here? Can you move it after the class?

File: lib/branch-list-view.coffee [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        console.log(data.toString())
+        @error = data.toString()
+      exit: (data) ->
+        code = parseInt(data)
+        unless code is 0
+          if @error.includes("error: Your local changes to the following files would be overwritten by checkout:")
+            options=
+              detail: "You must stash or commit your changes before attempting\n to checkout a new branch"
+              buttons:
+                [{text: 'Checkout Anyways',
```
Comment (benogle) [inline lib/branch-list-view.coffee]: This is pretty gnarly, is there anyway to clean this up, and make it more straightforward?

File: lib/branch-list-view.coffee [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+                    args: ['stash']
+                    cwd: cwd
+                    exit: (code) ->
+                      git
+                        args: ['checkout', branch.full]
+                        cwd: cwd
+                        exit: (code) ->
+                          emitter.emit 'on-change-branch', branch.name}]
+              dismissable: true
+            notification = atom.notifications.addError("You have uncommitted changes", options)
```
Comment (benogle) [inline lib/branch-list-view.coffee]: Use `addWarning`

File: lib/branch-list-view.coffee [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+      args: ["checkout", branch.full]
+      cwd: cwd
+      stderr: (data) ->
+        console.log(data.toString())
+        @error = data.toString()
+      exit: (data) ->
+        code = parseInt(data)
+        unless code is 0
+          if @error.includes("error: Your local changes to the following files would be overwritten by checkout:")
+            options=
```
Comment (benogle) [inline lib/branch-list-view.coffee]: space after `options`

**Comment by benogle** [comment]
Overall, this makes me nervous. We haven't done this yet because there are probably a lot of edge cases when someone has uncommitted changes. Does stashing cover every case? This kind of change has the opportunity to create a bunch of pain by losing work.

**Comment by TShapinsky** [comment]
@benogle what if it was just a dialog box telling you to commit your changes. I agree that the box should just be a warning, I was toying around trying to get the notification to stay on screen and forgot to change it back. Are there any other things I should be looking for when fixing my PR?

**Comment by squilter** [comment]
I like the idea of a notification box that just asks the user to manually commit uncommited changes and try again.

**Comment by damieng** [comment]
There has been no activity on this in well over a year so I'm going to close it down. Please feel free to discuss if it's worth continuing with.

**Pull request closed without merging**